### PR TITLE
Added boundary Coefficient classes based on GridFunction  [najlkin:pr8]

### PR DESCRIPTION
**Pull request by najlkin**
The set of the methods for obtaining the values from GridFunction is completed and new for boundary elements and faces are added. There are also new Coefficient classes utilizing these methods providing an easy access to the boundary/trace values for construction of the boundary conditions for example. Moreover, const qualifiers were added, since coefficients need read access to the grid functions only.

| PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#716](https://github.com/mfem/mfem/pull/716) | @najlkin | @v-dobrev | @tzanio + @mlstowell | 12/04/19 | ⌛due Invalid date | ⌛due Invalid date | |

### CODE REVIEW

**Comment by najlkin** [comment]
Please note that the bugfix #715 should be applied before/with this pull request, because usage of the boundary trace coefficient leads exactly to the problematic scenario mentioned there.

**Comment by najlkin** [comment]
Correct me if I am wrong, but the dimension of vector finite elements is given by the dimension of the reference space, i.e. `FiniteElement::GetDim()`, and not by the physical dimension of the space given by `Mesh::SpaceDimension()`. I have checked `VectorFiniteElement` and the size of the vector shape functions is like that. It seems that these two are confused at many places in the code. Should I fix it in the whole file or just in the committed part and make a separate issue/pull request?

**Comment by v-dobrev** [comment]
@najlkin, when a vector FE is mapped to physical space of larger dimension, e.g. H(curl)-conforming elements on a surface mesh, it has as many components as the physical space. For the case of H(curl)-conforming elements on a surface the mapped vector field is tangential to the surface and generally has 3 components. The way such elements are mapped is described briefly here:
https://github.com/mfem/mfem/blob/edfb62d8c27061a0de2b495eceffcee266b3d2cb/fem/fe.hpp#L253-L279

Note that the Jacobian matrix `J` is `sdim x dim` where `sdim` is `mesh.SpaceDimension()` and `dim` is `mesh.Dimension()`.

**Comment by najlkin** [comment]
@v-dobrev  Yes, you are right. The vector shape functions are internally calculated in the reference space in `CalcVShape(const IntegrationPoint& ...)` and transformed to the physical space in the call of `CalcVShape(ElementTransformation& ...)`, that puzzled me a bit. Ok, it means that everything should be alright even for the boundary elements then.

**Comment by v-dobrev** [comment]
The following branch is related to this PR: https://github.com/mfem/mfem/tree/grid-func-coeffs-on-bdr. Maybe something similar can be added here (allowing `GridFunctionCoefficient` and `VectorGridFunctionCoefficient` to be used with boundary element transformations).

**Comment by najlkin** [comment]
Yes, I could go that way and minimize the number of classes making the existent more "smart" by the test `mesh->Dimension() == T.GetDimension()` from your commit db3c027d1, if you think it is reliable.  In principle, I can maybe remove even the class `GridFunctionBdrTraceCoefficient` when I check if the finite element space has the boundary elements or not, but I do not know if it is a good idea, because the user maybe should stay aware of that L2 elements are not point-wise defined at the boundaries, but only in terms of the traces.

**Comment by najlkin** [comment]
Okay, I have moved everything to `GridFunctionCoefficient` and `VectorGridFunctionCoefficient` making the use more comfortable and being less pedantic on the user. I tried to explain the behavior of the coefficients in their descriptions, but if you do not like the idea of integration `GridFunctionBdrTraceCoefficient`and `VectorGridFunctionBdrTraceCoefficient` into the general ones then just revert the last commit. Anyway, the code fully surpasses db3c027 by the functionality at the moment.

**Comment by v-dobrev** [comment]
Thanks @najlkin, I think this approach makes it simpler for users, so I like it better.

File: fem/gridfunc.hpp [hunk 1]
```diff
@@ -141,24 +141,70 @@ class GridFunction : public Vector
 
    void GetVectorValue(int i, const IntegrationPoint &ip, Vector &val) const;
 
+   double GetBdrValue(int i, const IntegrationPoint &ir, int vdim = 1) const;
```
Comment (tzanio) [inline fem/gridfunc.hpp]: Can you add short comments to help users understand the difference between e.g. `GetBdrValue` and `GetBdrFaceValue`?

Along these lines, is there a way to consolidate some of these methods?

File: fem/gridfunc.cpp [hunk 2]
```diff
@@ -431,6 +431,186 @@ void GridFunction::GetVectorValue(int i, const IntegrationPoint &ip,
...
+      Tr->SetIntPoint(&ip);
+      be->CalcVShape(*Tr, vshape);
+      val.SetSize(spaceDim);
+      vshape.MultTranspose(loc_data, val);
+   }
+}
+
+double GridFunction::GetFaceValue(int i, int side, const IntegrationPoint &ip,
+                                  int vdim, int *pdir) const
+{
```
Comment (tzanio) [inline fem/gridfunc.cpp]: There seem to be a lot of code duplication in `GetFaceValue`, `GetFaceVectorValue`,  and the two `GetFaceValues`. Is there a way to consolidate it?

**Comment by tzanio** [comment]
A side comment: can you please open future PRs from branches inside the central mfem repository as opposed to your fork? This will simplify our workflow.

File: fem/gridfunc.cpp (hunk 2 above)
Comment (najlkin) [reply fem/gridfunc.cpp]: Yes, it is just about the choice of the face, but I introduced an auxiliary private function for that in 3b7ff80 replacing this procedure even in the existing methods.

**Comment by tzanio** [comment]
@najlkin: 3b7ff80 is better, thank you

Can you also address https://github.com/mfem/mfem/pull/716#discussion_r355092607 and update `CHANGELOG` please.

**Comment by mlstowell** [comment]
Hello, @najlkin,
I happened to try this fix for a DG field and it looks like there's an issue.  I put together a modified version of example 14 which demonstrates the problem (see below).  It looks like the element number in the transformation object is not a valid boundary element index.  I'm not sure how to resolve this but I'll help if I can.
Best wishes,
Mark
```
//                       MFEM Example 14 - Parallel Version
//
// Compile with: make ex14p
//
// Sample runs:  mpirun -np 4 ex14p -m ../data/inline-quad.mesh -o 0
//               mpirun -np 4 ex14p -m ../data/star.mesh -o 2
//               mpirun -np 4 ex14p -m ../data/star-mixed.mesh -o 2
//               mpirun -np 4 ex14p -m ../data/escher.mesh -s 1
//               mpirun -np 4 ex14p -m ../data/fichera.mesh -s 1 -k 1
//               mpirun -np 4 ex14p -m ../data/fichera-mixed.mesh -s 1 -k 1
//               mpirun -np 4 ex14p -m ../data/square-disc-p2.vtk -o 2
//               mpirun -np 4 ex14p -m ../data/square-disc-p3.mesh -o 3
//               mpirun -np 4 ex14p -m ../data/square-disc-nurbs.mesh -o 1
//               mpirun -np 4 ex14p -m ../data/disc-nurbs.mesh -rs 4 -o 2 -s 1 -k 0
//               mpirun -np 4 ex14p -m ../data/pipe-nurbs.mesh -o 1
//               mpirun -np 4 ex14p -m ../data/inline-segment.mesh -rs 5
//               mpirun -np 4 ex14p -m ../data/amr-quad.mesh -rs 3
//               mpirun -np 4 ex14p -m ../data/amr-hex.mesh
//
// Description:  This example code demonstrates the use of MFEM to define a
//               discontinuous Galerkin (DG) finite element discretization of
//               the Laplace problem -Delta u = 1 with homogeneous Dirichlet
//               boundary conditions. Finite element spaces of any order,
//               including zero on regular grids, are supported. The example
//               highlights the use of discontinuous spaces and DG-specific face
//               integrators.
//
//               We recommend viewing examples 1 and 9 before viewing this
//               example.

#include "mfem.hpp"
#include <fstream>
#include <iostream>

using namespace std;
using namespace mfem;

int main(int argc, char *argv[])
{
   // 1. Initialize MPI.
   int num_procs, myid;
   MPI_Init(&argc, &argv);
   MPI_Comm_size(MPI_COMM_WORLD, &num_procs);
   MPI_Comm_rank(MPI_COMM_WORLD, &myid);

   // 2. Parse command-line options.
   const char *mesh_file = "../data/star.mesh";
   int ser_ref_levels = -1;
   int par_ref_levels = 2;
   int order = 1;
   double sigma = -1.0;
   double kappa = -1.0;
   bool visualization = 1;

   bool gf_coef = false;

   OptionsParser args(argc, argv);
   args.AddOption(&mesh_file, "-m", "--mesh",
                  "Mesh file to use.");
   args.AddOption(&ser_ref_levels, "-rs", "--refine-serial",
                  "Number of times to refine the mesh uniformly in serial,"
                  " -1 for auto.");
   args.AddOption(&par_ref_levels, "-rp", "--refine-parallel",
                  "Number of times to refine the mesh uniformly in parallel.");
   args.AddOption(&order, "-o", "--order",
                  "Finite element order (polynomial degree) >= 0.");
   args.AddOption(&sigma, "-s", "--sigma",
                  "One of the two DG penalty parameters, typically +1/-1."
                  " See the documentation of class DGDiffusionIntegrator.");
   args.AddOption(&kappa, "-k", "--kappa",
                  "One of the two DG penalty parameters, should be positive."
                  " Negative values are replaced with (order+1)^2.");
   args.AddOption(&visualization, "-vis", "--visualization", "-no-vis",
                  "--no-visualization",
                  "Enable or disable GLVis visualization.");
   args.AddOption(&gf_coef, "-gfc", "--gridfunction-coef", "-no-gfc",
                  "--no-gridfunction-coef",
                  "Enable or disable GridFunctionCoefficient.");
   args.Parse();
   if (!args.Good())
   {
      if (myid == 0)
      {
         args.PrintUsage(cout);
      }
      MPI_Finalize();
      return 1;
   }
   if (kappa < 0)
   {
      kappa = (order+1)*(order+1);
   }
   if (myid == 0)
   {
      args.PrintOptions(cout);
   }

   // 3. Read the (serial) mesh from the given mesh file on all processors. We
   //    can handle triangular, quadrilateral, tetrahedral and hexahedral meshes
   //    with the same code. NURBS meshes are projected to second order meshes.
   Mesh *mesh = new Mesh(mesh_file, 1, 1);
   int dim = mesh->Dimension();

   // 4. Refine the serial mesh on all processors to increase the resolution. In
   //    this example we do 'ser_ref_levels' of uniform refinement. By default,
   //    or if ser_ref_levels < 0, we choose it to be the largest number that
   //    gives a final mesh with no more than 50,000 elements.
   {
      if (ser_ref_levels < 0)
      {
         ser_ref_levels = (int)floor(log(10000./mesh->GetNE())/log(2.)/dim);
      }
      for (int l = 0; l < ser_ref_levels; l++)
      {
         mesh->UniformRefinement();
      }
   }
   if (mesh->NURBSext)
   {
      mesh->SetCurvature(max(order, 1));
   }

   // 5. Define a parallel mesh by a partitioning of the serial mesh. Refine
   //    this mesh further in parallel to increase the resolution. Once the
   //    parallel mesh is defined, the serial mesh can be deleted.
   ParMesh *pmesh = new ParMesh(MPI_COMM_WORLD, *mesh);
   delete mesh;
   {
      for (int l = 0; l < par_ref_levels; l++)
      {
         pmesh->UniformRefinement();
      }
   }

   // 6. Define a parallel finite element space on the parallel mesh. Here we
   //    use discontinuous finite elements of the specified order >= 0.
   FiniteElementCollection *fec = new DG_FECollection(order, dim);
   ParFiniteElementSpace *fespace = new ParFiniteElementSpace(pmesh, fec);
   HYPRE_Int size = fespace->GlobalTrueVSize();
   if (myid == 0)
   {
      cout << "Number of unknowns: " << size << endl;
   }

   // 7. Set up the parallel linear form b(.) which corresponds to the
   //    right-hand side of the FEM linear system.
   ParLinearForm *b = new ParLinearForm(fespace);
   ConstantCoefficient one(1.0);
   ConstantCoefficient zero(0.0);
   b->AddDomainIntegrator(new DomainLFIntegrator(one));
   b->AddBdrFaceIntegrator(
      new DGDirichletLFIntegrator(one, one, sigma, kappa));
   b->Assemble();

   // 8. Define the solution vector x as a parallel finite element grid function
   //    corresponding to fespace. Initialize x with initial guess of zero.
   ParGridFunction x(fespace);
   x = 0.0;

   // 9. Set up the bilinear form a(.,.) on the finite element space
   //    corresponding to the Laplacian operator -Delta, by adding the Diffusion
   //    domain integrator and the interior and boundary DG face integrators.
   //    Note that boundary conditions are imposed weakly in the form, so there
   //    is no need for dof elimination. After serial and parallel assembly we
   //    extract the corresponding parallel matrix A.
   ParBilinearForm *a = new ParBilinearForm(fespace);
   a->AddDomainIntegrator(new DiffusionIntegrator(one));
   a->AddInteriorFaceIntegrator(new DGDiffusionIntegrator(one, sigma, kappa));
   a->AddBdrFaceIntegrator(new DGDiffusionIntegrator(one, sigma, kappa));
   a->Assemble();
   a->Finalize();

   // 10. Define the parallel (hypre) matrix and vectors representing a(.,.),
   //     b(.) and the finite element approximation.
   HypreParMatrix *A = a->ParallelAssemble();
   HypreParVector *B = b->ParallelAssemble();
   HypreParVector *X = x.ParallelProject();

   delete a;
   delete b;

   // 11. Depending on the symmetry of A, define and apply a parallel PCG or
   //     GMRES solver for AX=B using the BoomerAMG preconditioner from hypre.
   HypreSolver *amg = new HypreBoomerAMG(*A);
   if (sigma == -1.0)
   {
      HyprePCG pcg(*A);
      pcg.SetTol(1e-12);
      pcg.SetMaxIter(200);
      pcg.SetPrintLevel(2);
      pcg.SetPreconditioner(*amg);
      pcg.Mult(*B, *X);
   }
   else
   {
      GMRESSolver gmres(MPI_COMM_WORLD);
      gmres.SetAbsTol(0.0);
      gmres.SetRelTol(1e-12);
      gmres.SetMaxIter(200);
      gmres.SetKDim(10);
      gmres.SetPrintLevel(1);
      gmres.SetOperator(*A);
      gmres.SetPreconditioner(*amg);
      gmres.Mult(*B, *X);
   }
   delete amg;

   // 12. Extract the parallel grid function corresponding to the finite element
   //     approximation X. This is the local solution on each processor.
   x = *X;

   {
      GridFunctionCoefficient xCoef(&x);

      Coefficient *coefPtr = NULL;
      if (gf_coef)
      {
         coefPtr = &xCoef;
      }
      else
      {
         coefPtr = &one;
      }

      ParBilinearForm mx_bdr(fespace);
      mx_bdr.AddBdrFaceIntegrator(new BoundaryMassIntegrator(*coefPtr));
      mx_bdr.Assemble();

      ParLinearForm mx2(fespace);
      mx_bdr.Mult(x, mx2);
      double val = mx2(x);
      if (myid == 0)
      {
         cout << "Integral of solution^"
              << (gf_coef ? 3 : 2) << " on the boundary: " << val << endl;
      }
   }

   // 13. Save the refined mesh and the solution in parallel. This output can
   //     be viewed later using GLVis: "glvis -np <np> -m mesh -g sol".
   {
      ostringstream mesh_name, sol_name;
      mesh_name << "mesh." << setfill('0') << setw(6) << myid;
      sol_name << "sol." << setfill('0') << setw(6) << myid;

      ofstream mesh_ofs(mesh_name.str().c_str());
      mesh_ofs.precision(8);
      pmesh->Print(mesh_ofs);

      ofstream sol_ofs(sol_name.str().c_str());
      sol_ofs.precision(8);
      x.Save(sol_ofs);
   }

   // 14. Send the solution by socket to a GLVis server.
   if (visualization)
   {
      char vishost[] = "localhost";
      int  visport   = 19916;
      socketstream sol_sock(vishost, visport);
      sol_sock << "parallel " << num_procs << " " << myid << "\n";
      sol_sock.precision(8);
      sol_sock << "solution\n" << *pmesh << x << flush;
   }

   // 15. Free the used memory.
   delete X;
   delete B;
   delete A;

   delete fespace;
   delete fec;
   delete pmesh;

   MPI_Finalize();

   return 0;
}
```

File: fem/gridfunc.hpp (hunk 1 above)
Comment (najlkin) [reply fem/gridfunc.hpp]: Sure, I added comments in 4d7147b and added a note to `CHANGELOG`. Some consolidation seems to be also possible to smaller extend, I will look at it.

File: fem/gridfunc.hpp [hunk 3]
```diff
@@ -47,6 +47,23 @@ class GridFunction : public Vector
...
+       @param i     Face number
+       @param side  Side of the face (2 - automatic, side of the element with
+                    lower attribute is preferred; 1 - second element, if exists;
+                    0 - first element)
+       @param ftr   Pointer to FaceElementTransformations provided by
+                    Mesh::GetFaceElementTransformations()
+       @param elem  Optionally include ElementTransformation to @a ftr
+       @returns Indication of the side used (0 - first element, 1 - second
+       element) */
+   int GetFaceElementTransformations(int i, int side,
```
Comment (tzanio) [inline fem/gridfunc.hpp]: Thank you, this is great!

File: fem/gridfunc.hpp (hunk 1 above)
Comment (najlkin) [reply fem/gridfunc.hpp]: I consolidated Get(Bdr)Value(s) methods in 5b5821d. It saves just a few lines, but it is cleaner this way.

**Comment by najlkin** [comment]
> The new functionality is great, but I'm wondering if the implementation can avoid duplication, and if we can present a simpler interface to the users so they are not confused by the 10 different `GetBdr*Value*` methods.

@tzanio , I also find the number of methods somewhat excessive, but I have just followed the existing interface, where there are these 6 variants of the methods:
- Get*Value
- Get*VectorValue
- Get*Values
- Get*Values - returning the evaluation points
- Get*VectorValues
- Get*VectorValues - returning the evaluation points

these are then implemented for different kinds of elements:
- elements - were already present
- boundary elements
- face elements - were already partially present
- boundary face elements

This gives 6 * 4 = 24 methods in total, which is quite a lot, but the question if this number of variants provided by the interface can be reduced is not that much on me. For example, I am not sure if it is really necessary to provide the methods returning the evaluation points or if they are used at all. It is also possible to combine boundary and boundary face elements (returning point value or trace value depending on the type of the element), but it would be then confusing why there are methods for interior face elements and not boundary face elements. On the other hand, these methods are used directly just rarely based on my experience and rather a convenient interface is provided by newly extended `(Vector)GridFunctionCoefficient` classes automatically switching between the kinds of the elements.

**Comment by najlkin** [comment]
@mlstowell , thanks for the report. The problem is that `AddBdrFaceIntegrator()` in combination with `BoundaryMassIntegrator` evaluates `GridFunctionCoefficient` on a face element transformation, but this is not what the coefficient is designed for. It assumes that the there is a boundary element transformation for which you want to evaluate the trace of the function. I do not see a way how this could be recognized automatically, but I may add an extra coefficient for that (something like `FaceGridFunctionCoefficient`), which would operate on faces (both interior and boundary). That is a good suggestion :wink: .

**Comment by najlkin** [comment]
Ok, there it is `(Vector)GridFunctionFaceCoefficient`. Try it with them @mlstowell , it should work now for you.

**Comment by mlstowell** [comment]
Unfortunately, I'm getting a segfault with no stack trace and I don't have access to my usual debugger for some reason.  It seems to be crashing during the call to GetValue from within GetFaceValue.

**Comment by mlstowell** [comment]
I was able to get my debugger working.  The trouble seems to be that during assembly of the `BilinearForm` a `FaceElementTransformation` object is obtained and passed to the `BilinearFormIntegrator`.  However, this object is not passed to the coefficient, `Trans.Face` is passed instead.  Then inside the `GridFunctionFaceCoefficient`, `GetFaceValue` is called and this requests the `FaceElementTransformation` object again which causes it to be repopulated i.e. overwriting the data that was put into the `FaceElementTransformation` in the `BilinearForm::Assemble()` method.

Perhaps the best solution is to offer a `Coefficient::Eval` method which takes a `FaceElementTransformation` object so that this can be passed to the `GridFunction::GetFaceValue` method.  I haven't thought this through completely, do you see any potential pitfalls?

**Comment by v-dobrev** [comment]
One simple solution is to add this line:
```c++
   tr->Face->ElementNo = BdrElemNo;
```
after the following line: https://github.com/mfem/mfem/blob/55dbce3d70dd91c04582b41c69221d984ce02cbb/mesh/mesh.cpp#L937

It seems natural to use the boundary element id here similar to how we already use the boundary attribute.

**Comment by mlstowell** [comment]
It was a good idea but I'm still getting a segfault and my debugger is down again so I can't be more specific.  I'll try again in the morning.

**Comment by najlkin** [comment]
The segfault comes from that the call of `Mesh::GetFaceElementTransformations()` invoked by `GridFunctionFaceCoefficient` unmasks retrieval of face element transformation passed to the integrator in `FaceElementTransformations` before. Therefore, the hack how to make the code working is to add mask 16 to the calls of `Mesh::GetFaceElementTransformations()` in `GridFunction::GetFaceElementTransformations()` to get the face element transformation again. However, this is not a fix, because it saves the day now, but when someone will integrate over interior faces, the problem is there again, because one of the `IntegrationPointTransformation`s will be missing. This is a conceptual problem of `Mesh::GetFaceElementTransformations()` that a pointer is returned, but pointing to an internal structure instead of newly allocated one. It makes MFEM fast, but there are then unforeseen implications and internal clashes when the method is used repeatedly. I have run into a similar problem before in #715 , where I fixed that by adding an additional internal structure `Mesh`, but the very same functions for retrieval of the transformations are/can be used this time, so I do not see any elegant solution for this. A certain way how to get around this could be that the user manually instructs the coefficients to load full `FaceElementTransformations`, but that is not elegant even by far. Any ideas?

**Comment by v-dobrev** [comment]
> It was a good idea but I'm still getting a segfault and my debugger is down again so I can't be more specific. I'll try again in the morning.

@mlstowell, was this about my solution or something else? I tried the fix I suggested with the code you posted above and there were no segfaults.

**Comment by mlstowell** [comment]
Hello, @najlkin ,
I have a proposal which might simplify things and reduce the number of methods needed in the `GridFunction` class.  Rather than relying on an integer index and trying to figure out what this integer means we could use the `ElementTransformation` object and its `ElementNo` member data.  To remove ambiguity we can add a new piece of member data, e.g. `ElementTransformation::ElementType`, which  would indicate, using an enumeration, whether the `ElementNo` refers to an element, a boundary element, a face index, or an edge index. Using these two pieces of information we will know precisely how to extract the necessary values from the `GridFunction` in most cases.

Another standard case is evaluating a Discontinuous Galerkin field on the faces of a mesh e.g. in the  `DGDirichletLFIntegrator`.  In this case we have a `FaceElementTransformation` object which we could use.  We can define a new set of eval methods with a signature:
```
double Coefficient::Eval(FaceElementTransformation &tr, int side, const IntegrationPoint &ip);
```
So that we don't need to risk overwriting the transformation and we don't need to recompute it.

The last case I can think of is evaluating a Discontinuous Galerkin field on the boundary elements of a mesh when we are not in a Discontinuous Galerkin context.  In this case `GridFunction::GetValue` will need to request the `FaceElementTransformation` object and call itself with the `ElementTransformation` for the appropriate side.  In this case it should be ok to recompute the `FaceElementTransformation` object because it isn't already being used elsewhere.  Unfortunately, we will need to accept the cost of computing this at each integration point but I don't think that can be avoided.

This scheme should result in two `GridFunction` methods (or four if we need the `IntegrationRule` methods), one requiring an `ElementTransformation` and one for `FaceElementTransformation`.  Additionally it will require the new `Coefficient::Eval` methods but I suspect we can implement these in the base classes and override them for the `GridFunctionCoefficient` classes.

Does this sound reasonable?

Best wishes,
Mark

**Comment by mlstowell** [comment]
@v-dobrev , I was referring to your suggestion.  As @najlkin pointed out the mask used to generate the `FaceElementTransformation` did not produce a fully formed object (tr.Face was NULL).  Consequently, the evaluation succeeded but trying to set the next integration point back out in `AssembleFaceMatrix` lead to a segfault.  Perhaps also adjusting the mask would have succeeded but I resisted this because we already had a fully formed `FaceElementTransformation` and I'd rather not recompute it at every quadrature point.

**Comment by v-dobrev** [comment]
@mlstowell, did you try my suggestion with the new `GridFunctionFaceCoefficient`? It is meant to work with the original code using the standard `GridFunctionCoefficient`.

With this one line addition, your original code seems to work fine and there is no need to add `GridFunctionFaceCoefficient`.

**Comment by mlstowell** [comment]
@v-dobrev , I had been using the `GridFunctionFaceCoefficient`.  I agree that it now runs without error if I use your suggestion with the`GridFunctionCoefficient`.

I want to put together a small unit test or two to verify that these are working as expected.  The `ex14p` test is a bit too easy since the boundary values are zero everywhere.

**Comment by najlkin** [comment]
@mlstowell thanks for suggestions.
Regarding the first one to include the type of the element to `ElementTransformation` it would be nice, but it is not an issue now, because we know what type of the element we have implicitly. `GridFunctionCoefficient` operates on elements and boundary elements and `GridFunctionFaceCoefficient` on face elements. The problem is that this piece of information is useless when we cannot use corresponding functions of `Mesh`, because there are internal collisions.

Regarding that `Coefficient::Eval()` it is an appealing idea, but consider that no `BilinearFormIntegrator` would work with that, so you would have to rewrite a large portion of integrators, which I do not see feasible and it goes far beyond the scope of this pull request.

Evaluating DGs on boundary elements was my original intention and `GridFunctionCoefficient` and corresponding `GridFunction::GetBdrFaceValue(s)` should work there fine.

@v-dobrev your code works with `GridFunctionCoefficient`. It is a bit hacky workaround, but it works here :wink: . A trouble is that it may break existing code somewhere else expecting face number in face element transformation and not boundary element number (the attribute is not so critical as it does not lead to segfaults eventually). Secondly, this will not solve the problem for interior faces.

Anyway, to summarize, we need to decide which way to proceed:
1. Implementing the workaround (with a risk of breaking something) and dropping support for boundary/interior faces through `GridFunctionFaceCoefficient`, so the only access to boundary faces would be through `GridFunctionCoefficient` evaluated on boundary elements
2. Implementing the workaround (with a risk of breaking something) and keeping `GridFunctionFaceCoefficient`, but with a warning for users to not use it in the scenario with `BilinearForm` and rather suggest them to use `GridFunctionCoefficient` for the boundaries (so there will be nothing for assembly on interior faces)
3. Not implementing the workaround and load full `FaceElementTransformations` structure in `GridFunction::GetFaceElementTransformations` (conditionally or not). This will lead to the collision, but the `FaceElementTransformations` will be overwritten by the same data (this is how it works for all other types of elements, but there are no masks fortunately :smile: )
4. Not implementing the workaround, keeping `GridFunctionFaceCoefficient` and try to look for a way how to prevent the internal collisions in MFEM. For example by passing the existing transformations as @mlstowell has suggested or allocating new transformations structures or something else, but it may lead to extensive changes in MFEM very easily.
5. Not implementing the workaround, dropping `GridFunctionFaceCoefficient` or keeping it with a warning to not use it in @mlstowell 's scenario with `BilinearForm`, so there would be no chance to make Mark's code working directly.

What you say? @tzanio , what is your opinion on that?

**Comment by v-dobrev** [comment]
@najlkin, I think we should go with 1. Note that my suggestion is not exactly a "workaround" -- it is closer to a "bugfix". The method `Mesh::GetBdrFaceTransformations` (where my fix is) is actually used only when looping over boundary elements, so using the boundary element index should be the right thing to expect there.

In `(Vector)GridFunctionCoefficient`, the case of interior faces is not handled at the moment but I don't think we have a use case for that. If we need to handle this in the future, I think @mlstowell's suggestion to add a `ElementType` field to the `ElementTransformation` will be a good way to support this case.

@mlstowell, @tzanio, what do you think?

**Comment by tzanio** [comment]
I will like to hear from @mlstowell first. Mark, will @v-dobrev suggestion work for you needs?

**Comment by mlstowell** [comment]
Hello, @najlkin ,
I just created a WIP PR #1194 to illustrate the idea I was trying to describe.  It includes a more complete test program that we can modify to test your implementation.

Regarding your thought that we would need to modify all of the integrators to use the new `Coefficient::Eval(FaceElementTransformations &...)` method.   I don't believe this is the case.  There are only a handful of integrators that are designed to be used for `Face` or `BdrFace` integrals and these all take `FaceElementTransformations` objects in their argument lists so they should be easy to find.  The bigger change would be in the `Coefficient` classes which call other coefficient objects such as the `SumCoefficient`.  All of these will need newly implemented `Eval` methods.

I welcome your comments on this and I encourage you to try the [`examples/test_gfc.cpp`](https://github.com/mfem/mfem/blob/gf-getvalue-dev/examples/test_gfc.cpp) test program in your implementation.

Best wishes,
Mark

**Comment by najlkin** [comment]
Hi @mlstowell , thanks for the effort you are putting into this.

I am certainly not against modification of `Coefficient` class and I welcome the initiative in this direction in #1194 . It is basically point 4. mentioned in the summary. However, we both agree that it requires extensive changes in MFEM, so the question is if we want to go that direction now in this pull request, i.e. surpass it by your #1194 or find a different way how to prevent the collisions as stated in point 4.

The test code is also great :wink: and I have tried that with @v-dobrev 's workaround/bugfix and it works for all cases except the interior faces (as expected), but the support for interior faces goes beyond the original scope of this pull request and we can address it later maybe as @v-dobrev said (by #1194 eventually). However, I am still bit worried that we will break something by the workaround/bugfix. Anyway, are you fine with this (1.) or one of other more "conservative" procedures for the moment @mlstowell  ?

**Comment by mlstowell** [comment]
Hi, @najlkin ,
Yes, I'm fine with approach 1.  My only concern is the addition of new methods to `GridFunction` which may need to be removed in the future because there is often resistance to such removals based on the desire to maintain backward compatibility.

Part of the reason I'm so interested in this PR is that I have a need to use `GridFunctionCoefficients` based on DG fields to compute coefficients for DG diffusion operators.  Consequently, I need to compute integrals on both interior and boundary faces.  I was hoping this PR would add this functionality but I can certainly add this in a subsequent PR.

Best wishes,
Mark

**Comment by v-dobrev** [comment]
Hi @najlkin,

We discussed this with @tzanio and @mlstowell, and we decided to look into this a bit more. @mlstowell will develop further the ideas of #1194 and I will take a look at an alternative approach which may be simpler. We will keep you posted.

**Comment by najlkin** [comment]
Thanks, doing it a clear way and preventing collisions in MFEM is the best approach in long term, when support for interior faces is requested especially.

**Comment by tzanio** [comment]
@v-dobrev and @mlstowell -- any updates on this discussion?

**Comment by mlstowell** [comment]
@tzanio , I think we're waiting for feedback on PR #1194 .

**Comment by tzanio** [comment]
Postponed until the discussion of  #1194 is complete.

**Comment by tzanio** [comment]
Closed in favor of #1194
